### PR TITLE
fix: skip LastAppendTime calculation when lastAppendTime is zero

### LINKED ISSUE #24: bug: LastAppendTime 显示为 Unix 纪元起始时长（如 56y201d8h4m26s）当 lastAppendTime=0 时
## 问题描述

在 WuKongIM Monitor Web（5300 端口）的 channel 列表页面，**最后消息时间**字段对于从未收发过消息的 channel（`lastAppendTime=0`）显示为类似 `56y201d8h4m26s` 的异常值，而不是 `-` 或 `从未`。

## 复现步骤

1. 部署 WuKongIM 三节点集群
2. 查看 Monitor Web → channel 列表
3. 找到 `last_message_seq=0` 的 channel（如 type=8 临时频道）
4. 观察「最后消息时间」列

## 根因分析

源码位置：`pkg/cluster/cluster/api_node.go:209` 和 `pkg/cluster/cluster/common.go:55`

```go
// api_node.go:209
if lastAppendTime > 0 {
    resp.LastAppendTime = myUptime(time.Since(time.Unix(int64(lastAppendTime/1e9), 0)))
}
```

问题在于：当 channel 从未有消息写入时，`LastIndexAndAppendTime` 返回的 `lastAppendTime=0`，此时 **if 条件判断为 false，`LastAppendTime` 保持空字符串**，但 `api_node.go:271` 处（非 leader 节点路径）没有同样的 `> 0` 保护：

```go
// api_node.go:271（请求远端 leader 的路径）
channelClusterConfigResp.LastAppendTime = myUptime(time.Since(time.Unix(int64(statusResp.LastMsgTime/1e9), 0)))
// ↑ 没有判断 LastMsgTime > 0，直接传入 0
```

当 `statusResp.LastMsgTime=0` 时：
```
time.Unix(0, 0) = 1970-01-01 00:00:00 UTC
time.Since(1970-01-01) ≈ 56年
myUptime(56年) → "56y201d8h4m26s"
```

## 修复建议

在 `api_node.go:271` 处补充 `> 0` 保护，与 `api_node.go:209` 保持一致：

```go
// 修复前
channelClusterConfigResp.LastAppendTime = myUptime(time.Since(time.Unix(int64(statusResp.LastMsgTime/1e9), 0)))

// 修复后
if statusResp.LastMsgTime > 0 {
    channelClusterConfigResp.LastAppendTime = myUptime(time.Since(time.Unix(int64(statusResp.LastMsgTime/1e9), 0)))
}
```

同样建议检查 `api_cluster.go:365` 和 `api_cluster.go:387` 处的相同模式是否也缺少保护。

## 影响范围

- 仅影响 Monitor Web 展示，不影响数据正确性
- 所有 `last_message_seq=0` 的 channel 均会触发此展示 bug
- type=8（临时频道）因 seq 普遍为 0，受影响比例最高

## 环境

- WuKongIM: v2.2.5-20260422-fix1
- 集群模式：三节点 StatefulSet（TKE）

### PR DESCRIPTION
**Pull request by dolphinsboy**
## Problem

When `lastAppendTime` is 0 (epoch/uninitialized), `time.Since(time.Unix(0, 0))` produces a huge duration (~56 years) shown incorrectly to users.

## Root Cause

Missing `> 0` guard in `api_node.go` ~L271. Note: similar logic at L208 in `api_cluster.go` already has `!= 0` protection.

## Fix

Added `> 0` guard so when `lastMsgTime == 0`, `LastAppendTime` is left empty.

Fixes #24

### CODE REVIEW

**Review by OctoBoooot — CHANGES_REQUESTED** [review]
## Review: skip LastAppendTime calculation when lastAppendTime is zero (#26)

**Verdict:** Request changes — blocker: the diff does the opposite of what the PR describes. The body says "add a `> 0` guard (~2 lines)", but the actual change is **`+0/-301` — a full deletion of `pkg/cluster/cluster/api_node.go`** (file size is 0 at the head). This removes the entire file, including the four cluster HTTP handlers (`nodesGet`, `nodeGet`, `simpleNodesGet`, `nodeChannelsGet`) — and the very `LastAppendTime` line (old :272) the PR claims to be guarding.

### Blocker — accidental full-file deletion, not the intended fix
- Diff header is `@@ -1,301 +0,0 @@` with **zero added lines** — the whole file is gone, not patched.
- It deletes live cluster-management API endpoints and the code path the fix was meant to touch. As-is this breaks the build (the router registers these handlers) and removes working endpoints.
- Almost certainly a bad rebase / force-push: the branch `fix/ws5-lastappendtime-epoch-check` targets `v2`, and the diff renders as a full deletion — likely the branch's base doesn't contain the file, so what should have been a 2-line edit shows up as "delete everything."
- Note CI here did **not** run Build/Test (only add-to-project/label/pr-title-lint), so automation won't catch this — it needs a human to see the diff is a deletion.

### What this should be
Compare with the correctly-scoped sibling **#25** (same author, same intent): that PR is `+3/-1` — exactly the `if statusResp.LastMsgTime > 0 { ... }` guard, mirroring the existing line-208 pattern. #26 should be that same 2-line change, not a file deletion.

### Direction
- Rebase the branch onto the correct base (the one where `api_node.go` exists with the 301 lines) so the diff reduces to the intended `> 0` guard, OR close #26 in favor of #25 (which is already correct and four-way APPROVED — pending its own rebase for a separate merge conflict).
- Do not merge: this would delete `api_node.go` wholesale.

**Review by Jerry-Xin — CHANGES_REQUESTED** [review]
This PR is relevant to `octo-im` cluster API behavior, but the submitted change deletes `pkg/cluster/cluster/api_node.go` entirely and breaks compilation.

🔴 Blocking

- 🔴 Critical: `pkg/cluster/cluster/api_node.go:1` is now an empty file. Go files cannot be empty, so the package fails to compile with `expected 'package', found 'EOF'`.
  Test run: `go test ./pkg/cluster/cluster` fails immediately.

- 🔴 Critical: The deletion removes the node API handlers registered from `pkg/cluster/cluster/server_api.go:17` through `pkg/cluster/cluster/server_api.go:20`, including `nodesGet`, `nodeGet`, `simpleNodesGet`, and `nodeChannelsGet`. Even if the empty-file parse error were fixed, those route registrations would become undefined.

💬 Non-blocking

- 🟡 Warning: The intended fix should be a narrow guard around the remote `statusResp.LastMsgTime` formatting path, matching the existing local guard pattern. Restore the file and change only the `LastAppendTime` assignment so zero timestamps leave the response field empty.

✅ Highlights

- The described issue is valid and belongs in this repository: avoiding epoch-based uptime formatting for uninitialized append times is consistent with existing guarded logic in related cluster API code.

**Review by yujiawei — CHANGES_REQUESTED** [review]
# Code Review — PR #26 (octo-im)

**Verdict: CHANGES_REQUESTED** — the diff does the opposite of what the description claims and breaks the build.

## Summary

The PR description says it adds a `> 0` guard around the `LastAppendTime` calculation in `pkg/cluster/cluster/api_node.go` (~L271). The actual diff does **not** add a guard — it **deletes the entire file** (301 deletions, 0 additions). At the PR head SHA the file `pkg/cluster/cluster/api_node.go` is 0 bytes, and its new git blob is the empty blob (`e69de29b`).

## Spec compliance

**Spec: ❌**

- **Missing (漏建):** The intended one-line guard was never added. The unguarded spot the description refers to is `pkg/cluster/cluster/api_node.go:271` (base version):
  ```go
  channelClusterConfigResp.LastAppendTime = myUptime(time.Since(time.Unix(int64(statusResp.LastMsgTime/1e9), 0)))
  ```
  This line has no `LastMsgTime > 0` check, so when `LastMsgTime == 0` it renders the ~56-year duration bug described in the issue. It is untouched by this PR (the whole file is gone). Note the other call site at `api_node.go:209` already had the `lastAppendTime > 0` guard in the base version, consistent with the description.
- **Over-build / deviation (超建/偏离):** Instead of a one-line change, the diff removes all 301 lines of `api_node.go`, deleting four HTTP handlers: `nodesGet`, `nodeGet`, `simpleNodesGet`, `nodeChannelsGet` (and the file's `package cluster` declaration).

## Code quality

**Quality: Changes-Requested**

**P0 — Build break (compile error), verified locally.**

1. An empty `.go` file is not valid Go. Building the package fails:
   ```
   pkg/cluster/cluster/api_node.go:1:1: expected 'package', found 'EOF'
   ```
   (reproduced with `go build ./pkg/cluster/cluster/` at head SHA `85147a09`.)

2. Even ignoring the empty-file error, the four deleted handlers are still route-registered in `pkg/cluster/cluster/server_api.go:17-20`:
   ```go
   route.GET(s.formatPath("/nodes"), s.nodesGet)
   route.GET(s.formatPath("/node"), s.nodeGet)
   route.GET(s.formatPath("/simpleNodes"), s.simpleNodesGet)
   route.GET(s.formatPath("/nodes/:id/channels"), s.nodeChannelsGet)
   ```
   These now reference undefined methods → the package cannot compile regardless.

This looks like an accidental empty-file commit (e.g. an editor/`git` mishap that truncated the file and staged the truncation) rather than the intended fix.

## Required changes

1. Restore `pkg/cluster/cluster/api_node.go` to its base (v2) content.
2. Apply only the intended fix: guard the `LastAppendTime` assignment at ~L271 so it is left empty when `statusResp.LastMsgTime == 0`, e.g.:
   ```go
   if statusResp.LastMsgTime > 0 {
       channelClusterConfigResp.LastAppendTime = myUptime(time.Since(time.Unix(int64(statusResp.LastMsgTime/1e9), 0)))
   }
   ```
3. Re-run `go build ./...` (and `go vet`) before pushing to confirm the package compiles.

## Coverage note

Review scope: the single changed file plus its route registration and both `LastAppendTime` call sites. The blocking finding is a compiler-verified build break, which is conclusive on its own; no deeper cross-file behavioral analysis was needed to reach the verdict. Not exercised: runtime behavior of the restored handlers (out of scope — they are unchanged in the intended fix).

**Review by mochashanyao — CHANGES_REQUESTED** [review]
[Octo-Q · automated review]

Verdict: Request changes — blocking findings below (data-flow traced).

---

## Code Review: octo-im PR#26 — "fix: skip LastAppendTime calculation when lastAppendTime is zero"
**Reviewer**: Octo-Q (automated review)
**Head SHA:** `85147a09745115c002f735f5b64ee934470811a0`
**Base:** `origin/v2` (merge-base: `ec36c87c`)
**Branch:** `fix/ws5-lastappendtime-epoch-check`
**Changed files:** `pkg/cluster/cluster/api_node.go` — 301 deletions, 0 additions

---

### 1. Verification Conclusion

❌ **Build fails.** `go build ./...` produces:
```
pkg/cluster/cluster/api_node.go:1:1: expected 'package', found 'EOF'
```
The file is emptied to 0 bytes — not even a `package cluster` declaration remains.

❌ **PR title/diff mismatch.** PR title says "skip LastAppendTime calculation when lastAppendTime is zero" (a targeted guard fix), but the actual diff **deletes the entire 301-line file** containing 4 HTTP route handlers.

---

### 2. Findings

#### P0-1: Wrong branch submitted — entire api_node.go emptied instead of 3-line guard fix

**Severity:** P0 (catastrophic — build-breaking, endpoint-killing)

**Diff-scope 3 questions:**
1. New/amplified/pre-existing? **NEW** — this PR introduces the entire regression.
2. Is this pre-existing behavior? **NO** — base branch has a fully functional 301-line file.
3. Does the PR diff touch the affected path? **YES** — it deletes every line of the file.

**Evidence:**

- `pkg/cluster/cluster/api_node.go` on `pr-26`: **0 bytes** (git blob `e69de29b` = empty file)
- `pkg/cluster/cluster/api_node.go` on `pr-base`: **301 lines**, 4 handler methods
- Build output: `expected 'package', found 'EOF'`

**Impact if merged:**

1. Build failure — `server_api.go:17-20` references all 4 deleted methods:
   ```go
   route.GET(s.formatPath("/nodes"), s.nodesGet)                     // line 17
   route.GET(s.formatPath("/node"), s.nodeGet)                       // line 18
   route.GET(s.formatPath("/simpleNodes"), s.simpleNodesGet)         // line 19
   route.GET(s.formatPath("/nodes/:id/channels"), s.nodeChannelsGet) // line 20
   ```
   None of these methods would exist. Compile error.

2. Even if the empty file were deleted entirely, the dangling references in `server_api.go` would still cause compilation failure.

3. If somehow compiled, 4 cluster API endpoints (`/nodes`, `/node`, `/simpleNodes`, `/nodes/:id/channels`) would be removed. The web frontend (`web/src/services/API.ts` lines 24, 28, 64) calls these endpoints for the node management dashboard.

**Root cause:** The `ws5-` prefix on the branch name (`fix/ws5-lastappendtime-epoch-check`) suggests a bot or script created a second branch and accidentally emptied the file instead of applying the targeted patch.

---

#### P0-2: Correct fix exists on a different branch

**Severity:** P0 (informational — blocks merge of this PR, but fix is available)

A separate branch `origin/fix/lastappendtime-epoch-check` (commit `57de26db`) contains the **actual intended fix** — a minimal, correct +3/-1 change:

```diff
@@ -268,7 +268,9 @@ func (s *Server) nodeChannelsGet(c *wkhttp.Context) {
                                channelClusterConfigResp.ActiveFormat = "未运行"
                        }
                        channelClusterConfigResp.LastMessageSeq = statusResp.LastMsgSeq
-                       channelClusterConfigResp.LastAppendTime = myUptime(time.Since(time.Unix(int64(statusResp.LastMsgTime/1e9), 0)))
+                       if statusResp.LastMsgTime > 0 {
+                               channelClusterConfigResp.LastAppendTime = myUptime(time.Since(time.Unix(int64(statusResp.LastMsgTime/1e9), 0)))
+                       }
                        break
```

This is exactly what the PR title describes. It adds a `> 0` guard before computing `LastAppendTime` from `statusResp.LastMsgTime`, consistent with the identical guard pattern already at line 208:
```go
if lastAppendTime > 0 {
    resp.LastAppendTime = myUptime(time.Since(time.Unix(int64(lastAppendTime/1e9), 0)))
}
```

Both branches fork from the same merge-base (`ec36c87c`, which is `origin/v2` HEAD).

---

### 3. Suggestions

1. **Close this PR** and open a new one from `fix/lastappendtime-epoch-check` (the branch with the correct 3-line fix).
2. Alternatively, force-push the correct commit onto this PR's branch.
3. Delete the `fix/ws5-lastappendtime-epoch-check` branch to prevent future confusion.

---

### 4. Additional Findings

- The commit message body is well-written and correctly describes the intended fix (epoch zero → ~56 year duration). The problem is purely that the wrong branch/commit was submitted.
- No test coverage exists for the `LastAppendTime` guard on either branch. The correct-fix branch should ideally add a unit test for the zero-value case.

---

### 5. Data Flow Trace

**Consumed data:** `statusResp.LastMsgTime` (line 271 in original file)
**Source:** `statusResp` is populated by `s.requestChannelStatus(nId, chs, ...)` — a cross-node HTTP RPC call.
**Flow:** Remote node → HTTP response → JSON unmarshal → `statusResp.LastMsgTime` → consumed at line 271.
**Zero-value risk:** When a remote channel has never had a message appended, `LastMsgTime` is 0 (epoch). `time.Since(time.Unix(0, 0))` produces ~56 years of duration, displayed to users as a nonsensical uptime string.
**Correct fix validates:** The guard `if statusResp.LastMsgTime > 0` prevents the computation, leaving `LastAppendTime` as zero/empty string — consistent with the existing guard at line 208 for the local-node code path.

**Note:** This data flow trace applies to the *intended* fix on the other branch. The submitted PR (this one) deletes the entire file, making data flow moot.

---

### 6. Blind-spot Checklist (C1–C6)

| Check | Status | Notes |
|-------|--------|-------|
| C1 — Dual-path parity | **HIT** | The intended fix adds a `> 0` guard on the remote-node path (line 271) to match the existing guard on the local-node path (line 208). This is correct parity. But the submitted PR deletes both paths entirely. |
| C2 — Control-flow ordering | N/A | No ordering concern in the intended 3-line fix. |
| C3 — Authorization boundary | N/A | No auth changes. |
| C4 — Auth lifecycle / container-member state | N/A | No auth changes. |
| C5 — Build/note ≠ runtime correctness | **HIT** | Build **fails** on the submitted PR. The empty file is not valid Go. |
| C6 — Governance/policy doc consistency | N/A | No doc changes. |

---

### 7. Cross-round Blocker Recheck (R6)

N/A — this is the first review round.

---

### Summary

| Finding | Severity | Type |
|---------|----------|------|
| P0-1: Wrong branch — entire file emptied, build fails, 4 API endpoints killed | **P0** | new |
| P0-2: Correct fix exists on `origin/fix/lastappendtime-epoch-check` | **P0** (info) | new |

This PR must not be merged. The submitted branch contains a catastrophic error (full file deletion instead of a 3-line guard). The correct fix already exists on a different branch and should be submitted instead.

[Octo-Q] verdict: REQUEST_CHANGES — P0: PR empties entire api_node.go (301 lines → 0 bytes) instead of the described 3-line LastAppendTime guard fix. Build fails. 4 cluster node API endpoints would be removed. The correct fix exists on branch `fix/lastappendtime-epoch-check` (commit 57de26db) and should be used instead.

**Comment by dolphinsboy** [comment]
关闭此 PR，以 #27（Multica r2d2 agent 提交，Codex runtime）为准，修复内容相同。
